segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/SegmentKeyCache.java

```diff
@@ -505,12 +449,50 @@ synchronized void update(UUID keyHash, long segmentOffset, int currentGeneration
             serializeCacheValue(segmentOffset, entryData, entryOffset);
 
             // Update the cache and stats.
-            SegmentKeyCache.this.cache.insert(key, entryData);
-            this.size = entryData.length;
+            storeInCache(new ByteArraySegment(entryData));
             this.generation = currentGeneration;
             this.highestOffset = Math.max(this.highestOffset, segmentOffset);
         }
 
+        /**
+         * Removes the contents of this entry from the cache, if anything was stored there in the first place. Invoking
+         * this method will cause {@link #storeInCache} to throw an {@link IllegalStateException} going forward.
+         *
+         * @return True if there was anything evicted, false otherwise.
+         */
+        synchronized boolean evict() {
+            int address = this.cacheAddress;
+            this.cacheAddress = EVICTED_ADDRESS;
+            if (address >= 0) {
+                SegmentKeyCache.this.cacheStorage.delete(address);
+                return true;
+            }
+
+            return false;
+        }
+
+        @GuardedBy("this")
+        private byte[] getFromCache() {
```

Comment:
`getFromCache` in line 400 is not being invoked under a lock.

Reply:
Good catch. I changed this method to be `synchronized`.